fvm/evm/emulator/state/stateDB.go

```diff
@@ -0,0 +1,450 @@
+	for addr := range addresses {
+		sortedAddresses = append(sortedAddresses, addr)
+	}
+	sort.Slice(sortedAddresses,
+		func(i, j int) bool {
+			return bytes.Compare(sortedAddresses[i][:], sortedAddresses[j][:]) < 0
+		})
+
+	// update accounts
+	for _, addr := range sortedAddresses {
+		// TODO check if address is
+		if db.HasSuicided(addr) {
+			err = db.baseView.DeleteAccount(addr)
+			if err != nil {
+				return wrapError(err)
+			}
+			continue
+		}
+		if db.IsCreated(addr) {
+			err = db.baseView.CreateAccount(
+				addr,
+				db.GetBalance(addr),
+				db.GetNonce(addr),
+				db.GetCode(addr),
+				db.GetCodeHash(addr),
+			)
+			if err != nil {
+				return wrapError(err)
+			}
+			continue
+		}
+		err = db.baseView.UpdateAccount(
+			addr,
+			db.GetBalance(addr),
+			db.GetNonce(addr),
+			db.GetCode(addr),
+			db.GetCodeHash(addr),
+		)
+		if err != nil {
+			return wrapError(err)
+		}
+	}
+
+	// sort slots
+	sortedSlots := make([]types.SlotAddress, 0, len(slots))
+	for slot := range slots {
+		sortedSlots = append(sortedSlots, slot)
+	}
+	sort.Slice(sortedSlots, func(i, j int) bool {
+		comp := bytes.Compare(sortedSlots[i].Address[:], sortedSlots[j].Address[:])
+		if comp == 0 {
+			return bytes.Compare(sortedSlots[i].Key[:], sortedSlots[j].Key[:]) < 0
+		}
+		return comp < 0
+	})
+
+	// update slots
+	for _, sk := range sortedSlots {
+		err = db.baseView.UpdateSlot(
+			sk,
+			db.GetState(sk.Address, sk.Key),
+		)
+		if err != nil {
+			return wrapError(err)
+		}
+	}
+
+	// don't purge views yet, people might call the logs etc
+	err = db.baseView.Commit()
+	if err != nil {
+		return wrapError(err)
+	}
+	return nil
+}
+
+// Prepare is a highlevel logic that sadly is considered to be part of the
+// stateDB interface and not on the layers above.
+// based on parameters that are passed it updates accesslists
+func (db *StateDB) Prepare(rules gethParams.Rules, sender, coinbase gethCommon.Address, dest *gethCommon.Address, precompiles []gethCommon.Address, txAccesses gethTypes.AccessList) {
+	if rules.IsBerlin {
+		// no need for mutation
+		db.AddAddressToAccessList(sender)
+
+		if dest != nil {
+			db.AddAddressToAccessList(*dest)
+			// If it's a create-tx, the destination will be added inside egethVM.create
+		}
+		for _, addr := range precompiles {
+			db.AddAddressToAccessList(addr)
+		}
+		for _, el := range txAccesses {
+			db.AddAddressToAccessList(el.Address)
+			for _, key := range el.StorageKeys {
+				db.AddSlotToAccessList(el.Address, key)
+			}
+		}
+		if rules.IsShanghai { // EIP-3651: warm coinbase
+			db.AddAddressToAccessList(coinbase)
+		}
+	}
+}
+
+// Error returns the memorized database failure occurred earlier.
+func (s *StateDB) Error() error {
+	return wrapError(s.cachedError)
+}
+
+func (db *StateDB) lastestView() *DeltaView {
+	return db.views[len(db.views)-1]
+}
+
+// set error captures the first non-nil error it is called with.
+func (s *StateDB) handleError(err error) {
```

Comment:
```suggestion
func (db *StateDB) handleError(err error) {
```
for consistency. The same for `(s *StateDB) Error() error `

Reply:
fixed at [5625564](https://github.com/onflow/flow-go/pull/5168/commits/5625564100658440a905d349e6ea92f6bb62c132)